packages/flame/lib/src/game/game_widget/game_widget.dart

```diff
@@ -39,6 +40,8 @@ class GameWidget<T extends Game> extends StatefulWidget {
   /// The text direction to be used in text elements in a game.
   final TextDirection? textDirection;
 
+  final LoadingWidgetBuilder? loadingProgressBuilder;
```

Comment:
What's the difference between this and `loadingBuilder`?

Reply:
In loadingBuilder you should setup subscription on stream manually. With this class everything is already done, you build function just have additional "message" parameter. Also the base class takes care on removing StreamBuilder from widget tree when onLoad is finished - to prevent tree rebuilding on new messages. And allows you to setup transition between loading screen and game widget.

Of course everything could be done manually, even with loadBuilder (except animated transition). A new separate class allows user to forget about concrete messaging implementation and just to build own widget. It also allows to define strict message type - this is more flexible than receiving strings and less dirty than receiving "dynamic"

Reply:
I get the idea about the `LoadingWidget` class, but now `GameWidget` will have two parameters: `loadingBuilder`, and `loadingProgressBuilder`, which I feel would be confusing to users. Can't they be combined into one?

Reply:
I feel the API would be much simpler if we could say: `GameWidget` has the `loadingBuilder` parameter, where you can pass any widget you want. But if you create that widget based on `LoadingWidget`, then it will also receive loading progress events from the game.

Reply:
> But if you create that widget based on LoadingWidget, then it will also receive loading progress events from the game.

The problem is that loadingProgressBuilder is not a builder of a widget, it is pure mutable class. This is necessary for make stream setup under hood. But if to refactor in to Widget-based class, it become immutable and all widget's parameters should be passed only to constructor. So this will force user to do unnecessary copy-paste of boilerplate code. That's why `loadingProgressBuilder` and `loadingBuilder` are not compatible in this way.

I thought that `loadingProgressBuilder` could replace `loadingBuilder`  at future because it provides same functionality. To make things easier for users, who won't have any progress tracking, I thing it would be possible to subclass a special LoadingWidgetBuilder class with only `loadingBuilder`-like function as parameter. 

So, we have two options here: to make this change breaking but with easy migration steps. Or to make it non-breaking, by keeping both builders.

Reply:
> This is necessary for make stream setup under hood. But if to refactor in to Widget-based class, it become immutable and all widget's parameters should be passed only to constructor.

I'm not sure I understand what the problem is. So, suppose there is a `LoadingProgressWidget`, which has a function `getProgressMessage(context)`, which finds `GameWidget` from the build context, retrieves the `game` object from there, then looks up the last progress message from the `Game`. The API could look something like this:
```dart
class MyLoadingScreenWidget extends LoadingProgressWidget<double> {
  @override Widget build(BuildContext context) {
    final message = getProgressMessage(context);
    return Center(child: Text(message.toString()));
  }
}

void main() {
  runApp(
    GameWidget(
      game: game,
      loadingBuilder: MyLoadingScreenWidget(),
    )
  );
}
```

The `Game` would then be responsible for driving rebuilds of the game widget using its `refreshWidget` method.

Reply:
Yes, this should work too. But only advantage of this approach is backward compatibility, I think. Here is the list of disadvantages I see: 

**First.**  You can't implement transition between loading screen and game widget after game is loaded. To do that, you should to pass something into `MyLoadingScreenWidget` constructor, but you can't, besause `Stack(children: stackedWidgets);` does not exists here. You can fix it by using factory finction instead passing widget itself, like this:  `loadingBuilder: (Widget gameWidget) => MyLoadingScreenWidget(gameWidget),`. But with this you will force all users to repeat this unnecessary code and implement necessary parameter in constructor. And, as far as I know, IDE won't generate constructor for you automatically. Whereas with `loadingProgressBuilder` things become much easier: all variables are set in GameWidget builder and you dont care aboun constructor. And IDE will notice you about functions needed to be reimplemented. 

**Second.**  You will need different messaging implementation for widgets and for components. `reportLoadingProgress` can send messages both to loading widget and to components with `ProgressListener` mixin. Also streams is universal solution to subscribe any listener on it, for example to write logs or to create in-game debugging console (with cheatcodes, of course (-: ). With `HasProgressNotifier` you have `messagingStream` getter which returns broadcast stream. You can do with it whatever you want. But with approach 
>The Game would then be responsible for driving rebuilds of the game widget using its refreshWidget method 

we going to do separate implementations for loading screen, for components, etc. 
So with streams I hope to solve many posiible problems at core level.

Reply:
> But only advantage of this approach is backward compatibility, I think.

Another big advantage here is code simplicity. A complicated code creates maintenance problems in the long term (or even short term), as it increases chances of interference when new code needs to be added into the `GameWidget` class. Ideally, it would be nice to have the new functionality tightly encapsulated within one or two classes or mixins, without affecting the rest of the code too much. But the current approach modifies `GameWidget` quite significantly, which means it's likely that future changes to `GameWidget` may run into conflict with this functionality.

> First. You can't implement transition between loading screen and game widget after game is loaded.

True, but this functionality is independent from displaying the loading progress itself. This is something that could be implemented separately, and even have a separate parameter in the `GameWidget`: say, `loadingBuilderTransition`.

> Second. You will need different messaging implementation for widgets and for components.

This is true. But, there are really no components that would want to listen to game loading progress events. Simply because at the time the game is loading, no components are alive yet. And for logging purposes, you could simply add the needed code into the function that emits those progress messages in the first place. It would be quite simpler than creating the stream, then subscribing to the stream and unsubscribing from the stream.

As for loading components themselves (as described in #1776), I don't think streams are necessary either. Because, again, there would be only one component (`LoaderComponent`) interested in those "events", and it could just go and query the progress itself.

I understand that coming from Flutter, it seems natural to have streams everywhere to send messages to widgets to inform them of something that's happening, and forcing them to rebuild that way. But this is a Flutter approach, where everything is static and needs to be prodded by an external force in order to change.

But Flame is not like that. Flame is more dynamic, every component lives its own life, and updates and re-renders on every tick. A `LoaderComponent` doesn't need to subscribe to any events or streams, it can simply read `child.progress` value and draw itself based on that value, and that's it.

Reply:
>But the current approach modifies GameWidget quite significantly
Can't agree with this: there is just additional parameter and several lines of code to work with it. The rest of changes are related to formatting code and there is really nothing significant.

>This is something that could be implemented separately, and even have a separate parameter in the GameWidget: say, loadingBuilderTransition.

And this will change GameWidget much more than current implementation, where all possible things are bundled into `LoadingWidgetBuilder` and game's mixin `HasProgressNotifier`. 
Unfortunately, there is no way to avoid modification of `GameWidget`, if we want to implement transition after loading screen. 

>As for loading components themselves

Oh, no, my proposition was not about to use it while loading component themselves. In general, I thinking about such causes: b

>For example, there could be a game where the player travels on a world map, encounters a monster, and then goes into special fighting UI -- the UI, and the monster assets need to be loaded at this moment. Or the player enters a dungeon, and now you need to load the dungeon visuals. Etc.

In such situation components loading will be initialized from Game level, so we would not call progress notifier from component which loads itself. It is planned to be called from game, when game triggers loading process for a set of additional components.


But well, I see your point and I think I can't offer something useful for fixing the issue in the way you describe 👌🏼